wayland-client/src/global.rs

```diff
@@ -0,0 +1,226 @@
+#[derive(Debug)]
+pub struct GlobalList {
+    registry: wl_registry::WlRegistry,
+}
+
+impl GlobalList {
+    /// Gets the list of globals available to the connection.
+    pub fn new(conn: &Connection) -> Result<Self, GlobalError> {
+        let display = conn.display();
+        let data = RegistryState { globals: Mutex::new(Vec::new()) };
+
+        // Initialize the registry using the lower level wayland-backend API.
+        let registry =
+            conn.send_request(&display, wl_display::Request::GetRegistry {}, Some(Arc::new(data)))?;
+        let registry = wl_registry::WlRegistry::from_id(conn, registry)?;
+
+        // Roundtrip to wait until the server has send all globals over.
+        conn.roundtrip()?;
+
+        Ok(Self { registry })
+    }
+
+    /// Returns the list of advertised globals.
+    ///
+    /// This list is only valid until the next time the events are processed from the connection.
+    pub fn globals(&self) -> Vec<Global> {
+        let state = self.registry.data::<RegistryState>().unwrap();
+        let guard = state.globals.lock().unwrap();
+        guard.clone()
+    }
+
+    /// Binds a global, returning a new protocol object associated with the global.
+    ///
+    /// The `version` specifies the range of versions that should be bound. This function will guarantee the
+    /// version of the returned protocol object is the lower of the maximum requested version and the advertised
+    /// version.
+    ///
+    /// If the lower bound of the `version` is less than the version advertised by the server, then
+    /// [`BindError::UnsupportedVersion`] is returned.
+    ///
+    /// ## Multi-instance/Device globals.
+    ///
+    /// This function is not intended to be used with globals that have multiple instances such as `wl_output`
+    /// and `wl_seat`. These types of globals need their own initialization mechanism because these
+    /// multi-instance globals may be removed at runtime.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the maximum requested version is greater than the known maximum version of
+    /// the interface. The known maximum version is determined by the code generated using wayland-scanner.
+    pub fn bind_one<I, State, U>(
+        &self,
+        qh: &QueueHandle<State>,
+        version: RangeInclusive<u32>,
+        udata: U,
+    ) -> Result<I, BindError>
+    where
+        I: Proxy + 'static,
+        State: Dispatch<I, U> + 'static,
+        U: Send + Sync + 'static,
+    {
+        let version_start = *version.start();
+        let version_end = *version.end();
+        let interface = I::interface();
+
+        if *version.end() > interface.version {
+            // This is a panic because it's a compile-time programmer error, not a runtime error.
+            panic!("Maximum version ({}) of {} was higher than the proxy's maximum version ({}); outdated wayland XML files?",
+                version.end(), interface.name, interface.version);
```

Comment:
Isn't it breaking forward compat? You can generate code for older version and then new gets released?

Reply:
This is from some sctk code. The intention is to prevent programmer error by requesting for example version 5 of zwp_linux_dmabuf_v1 when there is only version 4 right now and then being shot in the foot when version 5 is actually released.

If this is regarding testing new protocol versions, you'll likely need to fork wayland-client/wayland-protocols anyways.

Reply:
I don't think this can break forwards compat, given this only panics if you're requesting a higher version than what wayland-rs supports. Updating wayland-rs to a higher supported version can thus never break code.